### DISCUSSION TITLE
Build the smart contract engine

### DISCUSSION THREAD

**Discussion by adrien-zinger** [massalabs/massa#1923] [Smart contract engine]
We want first to create an SC simulator, mocking the blockchain and the real massa ledger.

The project [wasm-simulation](https://github.com/massa/sc-runtime) that help us to find the best way to add a smart contract engine in massa.

(Go to the `brainstorming` branch to see the questions)

**Comment by adrien-zinger** [comment]
Those questions were pulled up:
- What does the LEDGER look like?
  - How do I know what's belong to me in the ledger User <-> [SC]
  - How do we know/pull an SC address? How do we create this kind of address?
- What is lacking here before merging with #1763?
- What subset of this code could be re-use as an execution engine in massa-node? all the `execution` lib!

Purpose of the main? 
Mock the blockchain behavior to be able to simulate SC wasm chunk without dealing with the blockchain.

**Comment by damip** [comment]
* the SCE ledger should basically be a map that maps Address to a SCELedgerData structure (see [this](https://github.com/massalabs/massa/pull/1763/files#diff-e515afc2f19989a188595956abae29d53f56a4ee836941979d0dd2b5bf6edb02R29-R33)) that contains:
  * balance: the SCE balance
  * bytecode: raw bytecode
  * data: a database associating a Hash to arbitrary data

**Comment by damip** [comment]
To generate a Smart Contract Address (without keys), you can for example hash( concat(operation.hash, index_of_this_smart_contract_among_those_created_by_this_operation)  )

**Comment by AureliaDolo** [comment]
> To generate a Smart Contract Address (without keys), you can for example hash( concat(operation.hash, index_of_this_smart_contract_among_those_created_by_this_operation) )

Should smart contract addresses and just addresses be the same rust struct ? In other worlds, if/when we implement #1105 should they have the same prefix ? They can share common methods but are they the same ?

**Comment by damip** [comment]
Maybe we can leave a bit somewhere in the ledger to say whether it is a SC address or a normal address... the only use case I see is being able to know whether the bytecode you're calling may change or is immutable
